lib/charms/opensearch/v0/opensearch_secrets.py

```diff
@@ -41,21 +41,63 @@
 logger = logging.getLogger(__name__)
 
 
+class SecretUpdateFailedError(Exception):
+    """The exception to be thrown when secret-changed fails to update its secret."""
+
+    pass
+
+
 class OpenSearchSecrets(Object, RelationDataStore):
     """Encapsulating Juju3 secrets handling."""
 
     LABEL_SEPARATOR = ":"
 
-    def __init__(self, charm, peer_relation: str):
+    def __init__(self, charm, peer_relation: str, secrets_to_cb: Dict[str, any] = {}):
         Object.__init__(self, charm, peer_relation)
         RelationDataStore.__init__(self, charm, peer_relation)
 
         self.cached_secrets = SecretCache()
+        self.secrets_to_cb = secrets_to_cb
 
         self.framework.observe(self._charm.on.secret_changed, self._on_secret_changed)
 
+    def _update_secret_value(self, event: SecretChangedEvent):
+        """Updates the secret value.
+
+        This method manages the secret update: compare the old value and
+        the new one (via peek=True). If they are different, then use the callback if
+        available and refreshes the secret. If the callback fails, then it raises an
+        exception to break the hook.
+
+        If the secret has a callback function, it will be called before the secret
+        is finally refreshed.
+        """
+        secret = event.secret
+        cb = self.secrets_to_cb.get(secret.label or secret.id)
+        scope = Scope.APP if Scope.APP in (secret.label or "") else Scope.UNIT
+        if not cb or (scope == Scope.APP and not self._charm.unit.is_leader()):
+            secret.get_content(refresh=True)
+            return
+        # There is a callback function to update the secret
+        old_val = secret.get_content()
+        new_val = secret.peek_content()
+        if new_val == old_val:
+            # Nothing to do, these dictionaries have the same content
+            return
+        if cb(old_val, new_val):
+            # Remove the old value
+            secret.get_content(refresh=True)
```

Comment:
NOTE: https://github.com/canonical/data-platform-libs/blob/main/lib/charms/data_platform_libs/v0/data_interfaces.py#L530-L542

If we want cross-versions backward compatibility, we may need 2 handle errors here